src/bin/shard/buddy.rs

```diff
@@ -340,3 +342,141 @@ pub fn find_name_accept_buddy(
         },
     )
 }
+
+pub fn buddy_warp(clients: &mut ClientMap, state: &mut ShardServerState) -> FFResult<()> {
+    let client = clients.get_self();
+    let pkt: sP_CL2FE_REQ_PC_BUDDY_WARP = *client.get_packet(P_CL2FE_REQ_PC_BUDDY_WARP)?;
+
+    let pc_id = client.get_player_id()?;
+    let player = state.get_player(pc_id)?;
+    let player_payzone_flag = player.flags.payzone_flag;
+    let player_skyway_state = player.skyway_ride.clone();
+    let buddy_uid = pkt.iBuddyPCUID;
+    if !player.is_buddies_with(buddy_uid) {
+        return Err(FFError::build(
+            Severity::Warning,
+            format!("{} is not buddies with player {}", player, buddy_uid),
+        ));
+    }
+
+    let mut invalid_warp = |error_code: i32, msg: String| -> FFResult<()> {
+        let response = sP_FE2CL_REP_PC_BUDDY_WARP_FAIL {
+            iBuddyPCUID: buddy_uid,
+            iErrorCode: error_code, // "Warp failure!\nUh oh! %s is in a location that you can't warp to! Try again later."
+        };
+        log_if_failed(
+            clients
+                .get_self()
+                .send_packet(P_FE2CL_REP_PC_BUDDY_WARP_FAIL, &response),
+        );
+        Err(FFError::build(Severity::Info, msg))
+    };
+
+    if player_skyway_state.is_some() {
+        return invalid_warp(
+            3,
+            format!(
+                "Cannot warp to buddy {} - player is currently on a skyway ride",
+                buddy_uid
+            ),
+        );
+    }
+
+    let search = PlayerSearchQuery::ByUID(buddy_uid);
+    let res = search.execute(state);
+    if res.is_none() {
+        if let Some(login_server) = clients.get_login_server() {
+            let cross_shard_pkt = sP_FE2LS_REQ_BUDDY_WARP_VALIDATION {
+                iWarpRequesterPCID: pc_id,
+                iWarpRequesterPCUID: player.get_uid(),
+                iWarpReceiverPCUID: buddy_uid,
+                iWarpRequesterPayzoneFlag: player_payzone_flag as i8,
+                iWarpRequesterShardID: config_get().shard.shard_id.get(),
```

Comment:
This field is redundant in the FE2LS packet since the login server will know the ID of the shard that sends this packet